value/eval.go

```diff
@@ -470,6 +470,178 @@ func ScanFirst(c Context, op string, v Value) Value {
 
 }
 
+// dataShape returns the data shape of v.
+// The data shape of a scalar is []int{}, to distinguish from a vector of length 1.
+func dataShape(v Value) []int {
+	switch v := v.(type) {
+	case Vector:
+		return []int{len(v)}
+	case *Matrix:
+		return v.Shape()
+	}
+	return []int{}
+}
+
+// TODO: When we update to Go 1.23, Edit ,s/iterSeq/iter.Seq[Value]/g
+type iterSeq[T any] func(yield func(T) bool)
```

Comment:
i asked for this not to happen now. not only are you adding for sequencing, you're adding generics to this codebase.

Reply:
It simplifies BinaryEach and Each considerably to abstract away the three choices about what's being iterated over from the code doing the iterating. My original implementation converted everything to Matrix and then computed a stride and passed subslices on each side, but that turned out to contain a subtle bug: `scalar foo@ v` turned the scalar left argument into a 1-element vector when invoking foo. This silently worked in my programs in 2021 but caused confusing errors in the new (better!) more-vector-preserving Ivy of 2024. To fix the accidental coercion, I need to handle the nine possible cases (scalar, vector, matrix) x (scalar, vector, matrix) with separate code. Cutting it into one function called twice with three cases is considerably nicer than attempting to spell them out.

I can keep the call sites that simple without generics (I only need `func(func(Value)bool)` not `iterSeq[Value]`) and with a slightly simpler callback function (I really only need `func(func(Value))`). Happy to do that if you want, but it seems strange to pretend not to know what the standard callback type will be.

Reply:
OK, but only if you promise to help me apply the approach universally. I do not like half updates.

Reply:
Will do.